### PR TITLE
luks_device: add allow_to_remove_last_key option

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Adds `allow_to_remove_last_key` option to `luks_device` as discussed [here](https://github.com/ansible/ansible/pull/48991#discussion_r255516337).

NOTE: this currently contains all commits of #52359 (integration tests). Will rebase once that PR is merged.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
luks_device

### CODE REVIEW

File: lib/ansible/modules/crypto/luks_device.py [hunk 1]
```diff
@@ -284,10 +295,41 @@ def run_luks_add_key(self, device, keyfile, new_keyfile):
...
+                raise ValueError('Error while dumping LUKS header from %s'
+                                 % (device, ))
+            keyslot_count = 0
+            keyslot_area = False
+            keyslot_re = re.compile(r'^Key Slot [0-9]+: ENABLED')
+            for line in result[STDOUT].splitlines():
+                if line.startswith('Keyslots:'):
+                    keyslot_area = True
+                elif line.startswith('  '):
+                    # New versions of "cryptsetup luksDump" use human-readable
```
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]: Turns out this is not a new version, but the LUKS2 header dump. (Learned that by looking at the `cryptsetup` sources...)

**Comment by felixfontein** [comment]
ready_for_review

**Comment by felixfontein** [comment]
@Xyon @puiterwijk since you wanted to have this check included, could you please take a look at this PR?

**Comment by felixfontein** [comment]
Rebased to current `devel` to get rid of stale_ci.

@japokorn @Xyon @puiterwijk I would really appreciate any kind of feedback for this one!

**Comment by felixfontein** [comment]
Restarting CI

**Comment by Xyon** [comment]
I don't (still) have much infra to test with, but the code LGTM - shipit

File: lib/ansible/modules/crypto/luks_device.py [hunk 2]
```diff
@@ -78,12 +81,20 @@
...
-              container. Data in there will be irreversibly lost
-              without a warning."
+            - "NOTE that removing keys is I(not idempotent). Trying to remove
+              a key which no longer exists results in an error."
+            - "NOTE that to remove the last key from a LUKS container, the
+              I(allow_to_remove_last_key) option must be set to C(yes)."
             - "BEWARE that working with keyfiles in plaintext is dangerous.
               Make sure that they are protected."
         type: path
+    allow_to_remove_last_key:
```
Comment (jamescassell) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
    force_remove_last_key:
```

File: lib/ansible/modules/crypto/luks_device.py [hunk 3]
```diff
@@ -78,12 +81,20 @@
...
+        description:
+            - "If set to C(yes), allows to remove the last key from a container."
```
Comment (jamescassell) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
            - "If set to C(yes), allows removing the last key from a container."
```

**Review by jamescassell — COMMENTED** [review]
General idea seems good.  Option needs a better name.

File: lib/ansible/modules/crypto/luks_device.py (hunk 2 above)
Comment (felixfontein) [reply lib/ansible/modules/crypto/luks_device.py]: Good suggestion! I changed the PR accordingly. @Xyon is the new name OK for you as well?

**Comment by felixfontein** [comment]
@Xyon @jamescassell if the current version is OK for you, could you please add a `shipit`?

**Review by jamescassell — APPROVED** [review]
shipit

**Comment by felixfontein** [comment]
@Xyon @jamescassell thanks for reviewing!
@gundalow thanks for merging!

**Pull request merged**